### PR TITLE
Add an option to inhibit suspend while playing

### LINKED ISSUE #5049: [Feature Request] Prevent monitor from sleeping when HDMI audio is used
If HDMI audio is used and the monitor goes to sleep, the audio is suspended until the monitor is awakened by the user.
Would it be possible to prevent the monitor to go to sleep, but only while Clementine is playing tracks?
Thus playback would not be interrupted every 10 minutes or so, depending on OS settings.

I think the current behaviour is standard, so the feature should be optional.

### PR DESCRIPTION
**Pull request by narunlifescience**
fixes #5049

### CODE REVIEW

File: src/ui/macidlehandler.cpp [hunk 1]
```diff
@@ -15,20 +16,43 @@
...
+  IOReturn reply = IOPMAssertionRelease(assertion_id_);
+
+  if (reply == kIOReturnSuccess) {
+    is_inhibit_ = false;
+  } else {
+    qLog(Warning) << "Failed to uninhibit screensaver/suspend";
+  }
+}
+
+bool WindowsIdleHandler::Isinhibited() {
```
Comment (hatstand) [inline src/ui/macidlehandler.cpp]: ?

File: src/ui/macidlehandler.h [hunk 2]
```diff
@@ -15,22 +16,24 @@
...
 
-  void Inhibit();
-  void Uninhibit();
+  void Inhibit(const char* reason) override;
+  void Uninhibit() override;
+  bool Isinhibited() override;
 
  private:
   IOPMAssertionID assertion_id_;
+  static bool is_inhibit_;
```
Comment (hatstand) [inline src/ui/macidlehandler.h]: Why is this static?

File: src/ui/idlehandler.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+*/
+
+#ifndef IDLEHANDLER_H
+#define IDLEHANDLER_H
+
+class IdleHandler {
+ public:
+  virtual ~IdleHandler() {}
+
+  static const char* kGnomeScreensaverService;
```
Comment (hatstand) [inline src/ui/idlehandler.h]: I think these can all be moved into `dbusidlehandler.cc` in an anonymous namespace

File: src/ui/mainwindow.cpp [hunk 4]
```diff
@@ -2417,6 +2427,11 @@ void MainWindow::EnsureSettingsDialogCreated() {
           SLOT(SetWiimotedevInterfaceActived(bool)));
 #endif
 
+  // Handle Suspend ststus
```
Comment (hatstand) [inline src/ui/mainwindow.cpp]: s/ststus/status/

File: src/ui/mainwindow.h [hunk 5]
```diff
@@ -373,6 +377,9 @@ class MainWindow : public QMainWindow, public PlatformInterface {
   PlayBehaviour menu_playmode_;
 
   BackgroundStreams* background_streams_;
+  IdleHandler* idlehandler_;
+  bool inhibit_suspend_while_playing_status_;
+  bool is_suspend_inhibited;
```
Comment (hatstand) [inline src/ui/mainwindow.h]: missing `_`

File: src/ui/playbacksettingspage.ui [hunk 6]
```diff
@@ -169,6 +178,22 @@
...
    <item>
+    <widget class="QGroupBox" name="power_management_groupBox">
+     <property name="title">
+      <string>System Power Management Options</string>
+     </property>
+     <layout class="QVBoxLayout" name="verticalLayout_4">
+      <item>
+       <widget class="QCheckBox" name="inhibit_suspend_while_playing">
+        <property name="text">
+         <string>Inhibit suspend while playing</string>
```
Comment (hatstand) [inline src/ui/playbacksettingspage.ui]: Can you write a more user-friendly description here?

File: src/ui/windowsidlehandler.h [hunk 7]
```diff
@@ -14,22 +15,24 @@
...
+  WindowsIdleHandler();
 
-  void Inhibit() override;
+  void Inhibit(const char*) override;
   void Uninhibit() override;
+  bool Isinhibited() override;
 
  private:
   EXECUTION_STATE previous_state_;
+  static bool is_inhibit_;
```
Comment (hatstand) [inline src/ui/windowsidlehandler.h]: Again, why static?

File: src/ui/macidlehandler.cpp (hunk 1 above)
Comment (narunlifescience) [inline src/ui/macidlehandler.cpp]: Oh! that looks nice :)

File: src/ui/macidlehandler.h (hunk 2 above)
Comment (narunlifescience) [inline src/ui/macidlehandler.h]: oops! let me change that..

File: src/ui/idlehandler.h (hunk 3 above)
Comment (narunlifescience) [inline src/ui/idlehandler.h]: hmm...

File: src/ui/mainwindow.cpp (hunk 4 above)
Comment (narunlifescience) [inline src/ui/mainwindow.cpp]: s

File: src/ui/mainwindow.h (hunk 5 above)
Comment (narunlifescience) [inline src/ui/mainwindow.h]: I was still playing around with the code.. probably missed _

File: src/ui/windowsidlehandler.h (hunk 7 above)
Comment (narunlifescience) [inline src/ui/windowsidlehandler.h]: I will change that..

File: src/ui/playbacksettingspage.ui (hunk 6 above)
Comment (narunlifescience) [inline src/ui/playbacksettingspage.ui]: OK

**Comment by narunlifescience** [comment]
Done!

**Comment by narunlifescience** [comment]
@hatstand any news on this PR ?

File: src/ui/dbusidlehandler.cc [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with Clementine.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+namespace {
+  
+  // Gnome screensaver 
+  const char* kGnomeScreensaverService
```
Comment (hatstand) [inline src/ui/dbusidlehandler.cc]: Sorry, these should be in dbusidlehandler.h &.cpp as members if they're required outside of that class or only in dbusidlehandler.cpp in an anonymous namespace if only used within that file.

**Comment by narunlifescience** [comment]
Done!

**Comment by amuttsch** [comment]
Breaks windows and mac build. Windows build can be easily fixed by correcting the `if (previous_state_! = 0) {` statements.

Mac I don't know how to get a `CfStringRef` from a `char *`.

File: src/ui/windowsidlehandler.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+          SetThreadExecutionState(ES_CONTINUOUS | ES_DISPLAY_REQUIRED);
+      break;
+    case IdleHandler::Suspend :
+      // resetting the system idle timer.
+      previous_state_ =
+          SetThreadExecutionState(ES_CONTINUOUS | ES_SYSTEM_REQUIRED);
+      break;
+  }
+
+  if (previous_state_! = 0) {
```
Comment (amuttsch) [inline src/ui/windowsidlehandler.cpp]: Fix `if`

File: src/ui/windowsidlehandler.cpp [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    is_inhibit_ = true;
+  } else {
+    qLog(Warning) << "Failed to inhibit screensaver/suspend";
+  }
+}
+
+void WindowsIdleHandler::Uninhibit() {
+  previous_state_ = 
+      SetThreadExecutionState(ES_CONTINUOUS | previous_state_);
+  if (previous_state_! = 0) {
```
Comment (amuttsch) [inline src/ui/windowsidlehandler.cpp]: Fix `if`

File: src/ui/macidlehandler.cpp [hunk 11]
```diff
@@ -15,20 +16,40 @@
...
+MacIdleHandler::MacIdleHandler() : assertion_id_(0), 
+                                   is_inhibit_(false) {}
 
-void MacScreensaver::Inhibit() {
-  IOPMAssertionCreateWithName(
+void MacIdleHandler::Inhibit(const char* reason) {
+  IOReturn reply = IOPMAssertionCreateWithName(
       kIOPMAssertPreventUserIdleDisplaySleep, kIOPMAssertionLevelOn,
-      CFSTR("Showing full-screen Clementine visualisations"), &assertion_id_);
+      CFSTR(reason), &assertion_id_);
```
Comment (amuttsch) [inline src/ui/macidlehandler.cpp]: `CFSTR` is a macro which cannot accept a `char *`. Check how to convert a `char *` to a `CFStringRef`.

File: src/ui/windowsidlehandler.cpp (hunk 9 above)
Comment (narunlifescience) [inline src/ui/windowsidlehandler.cpp]: Sorry! will fix it..

File: src/ui/windowsidlehandler.cpp (hunk 10 above)
Comment (narunlifescience) [inline src/ui/windowsidlehandler.cpp]: will fix it.

File: src/ui/macidlehandler.cpp (hunk 11 above)
Comment (narunlifescience) [inline src/ui/macidlehandler.cpp]: I think CFStringCreateWithCString function can do that..

**Comment by jtlindsey** [comment]
Is there any work around to allow clementine to prevent sleep? I'm using Ubuntu 16.04LTS

**Comment by narunlifescience** [comment]
@hatstand could you please check #5202 when you are free.. if you find it OK, I will try to resolve the conflicts or make a new pull request..

**Pull request merged**